File: 2017/day07.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day07.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import parsefile, Merge

nodes = parsefile(file_name, [[str, 1, str, 1, None, 1, Merge([str, ","]), 0], "\n"])


class Node:
	def __init__(self, name, weight):
		self.name = name
		self.weight = weight
		self.child_nodes = []
		self.parent_node = None

	def calc_weight(self):
		s = self.weight
		for c in self.child_nodes:
			s += c.calc_weight()

		return s

	def is_balanced(self):
		if len(self.child_nodes) == 0:
			return True, 0

		counts = {}
		for c in self.child_nodes:
			res = c.is_balanced()
			if not res[0]:
				return res

			cw = c.calc_weight()
			if cw in counts:
				counts[cw] += 1
			else:
				counts[cw] = 1

		if len(counts) == 1:
			return True, 0

		if len(counts) > 2:
			return False, -1

		sl = sorted(counts.items(), key=lambda x: x[1])
		diff = sl[1][0] - sl[0][0]

		tn = None
		for c in self.child_nodes:
			if c.calc_weight() == sl[0][0]:
				tn = c
				break

		return False, tn.weight + diff


node_dict = {}

for n in nodes:
	w = int(n[1][1:-1])
	node = Node(n[0], w)
	node_dict[n[0]] = node

# build tree
for n in nodes:
	node = node_dict[n[0]]
	for c in n[2:]:
		c_node = node_dict[c]
		node.child_nodes.append(c_node)
		c_node.parent_node = node

tree_node = None


def part1():
	global tree_node
	for n, node in node_dict.items():
		if node.parent_node is None:
			tree_node = node
			return n


def part2():
	res = tree_node.is_balanced()
	return res[1]


p1()
p2()

File: cpp/2017/day07.cpp
#include "../aocHelper.h"


class Day07 : public BaseDay
{
public:
	Day07() : BaseDay("07") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		struct Node
		{
			string name;
			int weight;
			vector<Node*> child_nodes;
			Node* parent_node = nullptr;

			Node(string& name, int weight) : name(name), weight(weight)
			{}

			int calc_weight()
			{
				int s = this->weight;
				for (auto& c : this->child_nodes)
				{
					s += c->calc_weight();
				}
				return s;
			}

			pair<bool, int> is_balanced()
			{
				if (this->child_nodes.size() == 0)
				{
					return { true, 0 };
				}

				unordered_map<int, int> counts;
				for (auto& c : this->child_nodes)
				{
					auto res = c->is_balanced();
					if (!res.first)
					{
						return res;
					}

					int cw = c->calc_weight();
					if (counts.contains(cw))
					{
						counts[cw]++;
					}
					else
					{
						counts[cw] = 1;
					}
				}

				if (counts.size() == 1)
				{
					return { true, 0 };
				}

				if (counts.size() > 2)
				{
					return { false, -1 };
				}

				vector<int> keys;
				for (auto& p : counts)
				{
					keys.push_back(p.first);
				}

				sort(keys.begin(), keys.end(), [&](const int& a, const int& b)
				{
					return counts[a] < counts[b];
				});

				int diff = keys[1] - keys[0];

				Node* tmp = nullptr;

				for (auto& c : this->child_nodes)
				{
					if (c->calc_weight() == keys[0])
					{
						tmp = c;
						break;
					}
				}

				return { false, tmp->weight + diff };
			}
		};

		struct raw_node
		{
			string name;
			int weight;
			vector<string> child_nodes;
		};

		vector<raw_node> raw_nodes;
		unordered_map<string, Node*> node_dict;

		// parse input

		string str;

		while (*input != '\0')
		{
			raw_node curr_node;

			while (*input >= 'a' && *input <= 'z')
			{
				curr_node.name += *input;
				input++;
			}

			input += 2; // slip " ("

			curr_node.weight = numericParse<int>(input);

			input++; // skip ')'

			if (*input != '\n')
			{
				input += 4; // skip " -> "

				string str;

				while (*input != '\n')
				{
					str += *input;
					input++;

					if (*input == ',')
					{
						input += 2; // skip ", "
						curr_node.child_nodes.push_back(str);
						str = "";
					}
				}

				curr_node.child_nodes.push_back(str);
			}

			raw_nodes.push_back(curr_node);
			node_dict[curr_node.name] = new Node(curr_node.name, curr_node.weight);

			input++; // skip '\n'
		}

		// build tree
		for (auto& rn : raw_nodes)
		{
			Node* node = node_dict[rn.name];

			for (auto& c : rn.child_nodes)
			{
				Node* child_node = node_dict[c];
				node->child_nodes.push_back(child_node);
				child_node->parent_node = node;
			}
		}

		Node* tree_node = nullptr;

		// part 1
		for (auto& p : node_dict)
		{
			if (p.second->parent_node == nullptr)
			{
				tree_node = p.second;
				memcpy(stringResult.first, p.first.c_str(), p.first.length());
				break;
			}
		}

		// part 2
		auto res = tree_node->is_balanced();
		part2 = res.second;

		for (auto& p : node_dict)
		{
			delete p.second;
		}

		return { part1, part2 };
	}
};
